### ISSUE TITLE
--define has gone missing in v8

### ISSUE THREAD

**Issue by nzdjb** [python-semantic-release/python-semantic-release#625]
## The problem

We use the --define argument (or, specifically, -D) to version in a templated build file to apply the same build settings to all of our hundreds of projects so that we don't need to maintain them in pyproject.toml. In v8, --define is no longer available.

I assume this was intentional, but it also doesn't appear in the v8.0.0 release notes.

On v7:
```shell
semantic-release --noop version -D version_toml=$PYPROJECT:$VERSION_VAR -D parser_angular_patch_types=fix,perf,chore,docs
Creating new version
Current version: 0.3.11, Current release version: 0.3.11
warning: No operation mode. Should have bumped from 0.3.11 to 0.4.0
```

On v8:
```shell
semantic-release --noop version -D version_toml=$PYPROJECT:$VERSION_VAR -D parser_angular_patch_types=fix,perf,chore,docs
🛡 You are running in no-operation mode, because the '--noop' flag was supplied
Usage: semantic-release [OPTIONS] COMMAND [ARGS]...
Try 'semantic-release -h' for help.

Error: Missing keys 'tool.semantic_release' or 'semantic_release' in pyproject.toml
```

## Expected behavior

I expected --define to be available in v8.

## Environment

```shell
python --version
Python 3.11.4
```

```shell
pip --version
pip 23.1.2 from /usr/local/lib/python3.11/site-packages/pip (python 3.11)
```

```shell
semantic-release --version
semantic-release, version 8.0.0
```

```shell
pip freeze
attrs==23.1.0      
build==0.10.0      
CacheControl==0.12.14
certifi==2023.5.7                             
cffi==1.15.1          
charset-normalizer==3.2.0                                                                                                                                                                                             
cleo==2.0.1
click==8.1.5
crashtest==0.4.1
cryptography==41.0.2
distlib==0.3.6
dotty-dict==1.3.1
dulwich==0.21.5
filelock==3.12.2
gitdb==4.0.10
GitPython==3.1.32
html5lib==1.1
idna==3.4
importlib-metadata==6.8.0
importlib-resources==5.7.0
installer==0.7.0
jaraco.classes==3.3.0
jeepney==0.8.0
Jinja2==3.1.2
jsonschema==4.18.3
jsonschema-specifications==2023.6.1
keyring==23.13.1
lockfile==0.12.2
markdown-it-py==3.0.0
MarkupSafe==2.1.3
mdurl==0.1.2
more-itertools==9.1.0
msgpack==1.0.5
packaging==23.1
pexpect==4.8.0
pkginfo==1.9.6
platformdirs==3.9.1
poetry==1.5.1
poetry-core==1.6.1
poetry-plugin-export==1.4.0
ptyprocess==0.7.0
pycparser==2.21
pydantic==1.10.11
Pygments==2.15.1
pyproject_hooks==1.0.0
python-gitlab==3.15.0
python-semantic-release==8.0.0
rapidfuzz==2.15.1
referencing==0.29.1
requests==2.31.0
requests-toolbelt==1.0.0
rich==13.4.2
rpds-py==0.8.10
SecretStorage==3.3.3
shellingham==1.5.0.post1
six==1.16.0
smmap==5.0.0
tomlkit==0.11.8
trove-classifiers==2023.7.6
typing_extensions==4.7.1
urllib3==1.26.16
virtualenv==20.24.0
webencodings==0.5.1
zipp==3.16.2
```

```shell
poetry --version
Poetry (version 1.5.1)
```

## Configuration

Intentionally no semantic-release configuration.

build-system configuration:
```toml
[build-system]
build-backend = "poetry.core.masonry.api"
```

## Logs

```shell
[03:10:47] DEBUG    [semantic_release.cli.commands.main] DEBUG main.main: logging level set to: DEBUG                                                                                                      main.py:106
🛡 You are running in no-operation mode, because the '--noop' flag was supplied
           DEBUG    [semantic_release.cli.commands.main] DEBUG main.main: global cli options: GlobalCommandLineOptions(noop=True, verbosity=2, config_file='pyproject.toml', strict=False)                 main.py:124
           INFO     [semantic_release.cli.commands.main] INFO main.main: Loading TOML configuration from pyproject.toml                                                                                    main.py:128
Usage: semantic-release [OPTIONS] COMMAND [ARGS]...
Try 'semantic-release -h' for help.

Error: Missing keys 'tool.semantic_release' or 'semantic_release' in pyproject.toml
```

**Comment by nzdjb** [comment]
I've now found [docs/migrating_from_v7.rst](https://github.com/python-semantic-release/python-semantic-release/blob/master/docs/migrating_from_v7.rst), which also makes no mention of --define being removed.

**Comment by nzdjb** [comment]
I see this has now been addressed in the migrating_from_v7.rst file.

I agree that --define was not the right solution and it's valid to remove it, however I disagree that excluding complex types from parameters is acceptable. It should be possible to configure the entire application via parameters (and ideally also environment variables, to make running the application in a container more pleasant).

Using --config is also not a particularly valid workaround as the options are to either write this into the workspace, which pollutes the workspace unnecessarily or to a location outside the workspace which makes the assumption that such a location exists and is writable.

Also, a lack of configuration (i.e. no [semantic_release] section) should not cause an error and should instead use all the defaults (the same as how an empty [semantic_release] section does).

The only things I need to get my builds moving on v8 are:
 * Fix the issue of requiring [semantic_release] section.
 * Allow configuration of patch_tags via command line.
 * Allow configuration of version_toml via command line.

I'll create issues and probably PRs for each of these.

**Comment by nzdjb** [comment]
Opened as #634, #635 and #636.

**Comment by nzdjb** [comment]
Thanks for the response.

> I agree with you that removal of `--define` could definitely have been called out more explicitly and was a big omission from the release docs - will hold my hands up and apologise, I thought I'd included it in the migration notes and obviously overlooked it.

It's often helpful to release a stable version with a warning that a feature will be removed and a link to a suggested replacement. This allows discussion and fixes to occur before the breaking change arrives. It also serves to allow deprecation notices to be put into release notes ahead of time.

I'm much more likely to see a warning in my build logs or a mention in the stable release notes than I am a change in the code or changelog of an unstable version.

> However I'm not sure that there's inherently an issue with having clear demarcations between what can be specified as envvars, configuration or on the command-line - it's in step with the Zen:
> 
> > There should be one-- and preferably only one --obvious way to do it.

The Zen applies specifically to the design of Python itself. Build tools should be unopinionated in how they accept input as they'll be used in all sorts of configurations and environments. It should be as easy as possible to fall into the Pit of Success, no matter the direction you're coming from.

FWIW, Click makes it very easy to support env vars with auto_envvar_prefix: https://click.palletsprojects.com/en/8.1.x/options/#values-from-environment-variables

> That said I do understand wanting to be able to invoke PSR entirely from the command-line.
> 
> I do think there are options - even for workspaces that aren't writable - for example `ConfigMaps` to add files to containers running on K8s, with a mounted directory in a `docker run`, or even process substitution (note I'm not the biggest fan of this, and it won't work at the moment - see below)

There are definitely options, though both ConfigMaps and mounted directories assume the user has access to configure the build infrastructure. Process substitution is not portable to all shells and also risks making build files overly complicated and opaque.

> I see the following as a path forward:
> 
> * I agree with [pyproject.toml without [tool.semantic_release] should use defaults instead of error. #634](https://github.com/python-semantic-release/python-semantic-release/issues/634) but need this to behave the same every configuration format, not just TOML. It will have a knock-on on any mandatory configuration options in the future, but I don't think that's a worry for the library as-is
> * I think looking at this issue, I would prefer PSR to try parsing explicitly-specified config with `tomlkit.load`, `json.load` and any other supported config formats before giving up, instead of just trying to guess based on file extension. This would also allow the process-substitution solution above. I've raised [fix: handle missing configuration #644](https://github.com/python-semantic-release/python-semantic-release/pull/644) to do this
> * Enhance the existing CLI with explicit options to match the configuration
> 
> What are your thoughts?

That would work for me. Thanks.
